Category Listing Screen - Viewmodel Implementation

AjeshRPai opened the pull request:
Parent #15550 

This PR adds the following 

#### Changes to `CategoriesListViewModel` 
- Expose a LiveData which represents the UI State of the Screen  
- Fetch Categories using `getCategoriesUseCase` using Site Model passed from UI 
- Adds a Sealed UI State class `UiState` which holds the Information of the State (Content,Error, Loading) as well as the data needed to be Displayed on the Screen 
- Adds the Logic of updating the Correct UI state in LiveData according to the Data present in the DB as well as the Callback from Taxonomy Store 
- Adds a dispatcher and the associated logic to subscribe for the `onTaxonomyChanged` events 

#### Changes to `CategoriesListViewModelTest`
- Adds the Unit Tests for the Logic in `CategoriesListViewModel` 😄 
 

#### Changes to CategoriesListFragment
- Adds the Logic of getting the `SiteModel` instance from Bundle 

#### Changes to strings.xml 
- Adds Title and Subtitle to be displayed on UI for the Error states `NoConnection` and `GenericError`  

## Regression Notes
1. Potential unintended areas of impact
None

2. What I did to test those areas of impact (or what existing automated tests I relied on)
N/A

3. What automated tests I added (or what prevented me from doing so)
Added Unit Tests for Viewmodel Changes 

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

AjeshRPai committed:
Refactor: CategoriesListViewModel, CategoriesListFragment and GetCategoriesUseCase

AjeshRPai committed:
Refactor: CategoriesListViewModelTest and CategoriesListViewModel

Adds the Tests in CategoriesListViewModelTest
Todo notation added for Changing the AnalyticsTrackerWrapper and Log tag
Adds Missing Return statement on the callback in case of error in CategoriesListViewModel

AjeshRPai committed:
Refactor: Use getSiteCategories from GetCategoriesUseCase

Changes the method that's being used to fetch the Categories of the Site
The current method returns the Categories with Parent Child Relationship

AjeshRPai committed:
Merge branch 'wordpress-mobile:develop' into issue/5382-viewmodel-implementation

AjeshRPai committed:
Adds Resource Strings for No Network

Adds String Resources for No Network state in CategoriesListViewModel while fetching Categories
Refactors Log Tag being used in case of Taxonomy Changed error

AjeshRPai committed:
Merge remote-tracking branch 'origin/issue/5382-viewmodel-implementation' into issue/5382-viewmodel-implementation

AjeshRPai committed:
Add: Resource Strings for Generic Error State

Add Generic Error message strings which will be shown if the request to fetch the Categories results in an error.

AjeshRPai committed:
Remove: Unused Dependency

Remove the Unused Dependency - AnalyticsTrackerWrapper from CategoriesListViewModel and tests.

AjeshRPai committed:
Refactor: The Default Dispatcher used in Viewmodel

Refactor: Change the Default Dispatcher to BG Dispatcher and pass it to the Base Scoped Viewmodel
All launch calls will work on the Default Dispatcher which in this case will be bgDispatcher.

AjeshRPai committed:
Remove: UI Changes and Unused state

Remove the UI Changes as those will be taken up later and pushed in another branch.
Remove the Unused Navigation state in CategoriesListViewModel

ashiagr commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/categories/CategoriesListViewModel.kt:
```diff
@@ -1,13 +1,131 @@
+                override val buttonText = UiStringRes(string.retry)
+            }
+
+            data class NoCategories(override val action: () -> Unit) : Error() {
```
Warning ⚠️: It seems we will not need `NoCategories` case considering all posts are grouped under "Uncategorized" by default in other places, and the default category cannot be deleted. Wdyt?

AjeshRPai replied:
Yeah, @ashiagr . You are correct 🙌  . We wouldn't need the `NoCategories ` case . I will remove this error state class. 👍 Shall I remove the associated string resources as well ?.

ashiagr replied:
> Shall I remove the associated string resources as well ?.

Not sure if I'm missing something (could be translation-related doubts which is why you asked this question) 😅, once we remove this case, all associated unused strings and tests should be removed as well.

AjeshRPai replied:
Removed the Class in https://github.com/wordpress-mobile/WordPress-Android/pull/15552/commits/e20873663dd9a7a108c9ef4d668ad21209d6e8f3

ashiagr commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/categories/CategoriesListViewModel.kt:
```diff
@@ -1,13 +1,131 @@
+
+    private fun fetchCategoriesFromNetwork(isInvokedFromInit: Boolean = false) {
+        launch {
+            if (networkUtilsWrapper.isNetworkAvailable()) {
+                if (!isInvokedFromInit)
+                    _uiState.postValue(Loading)
+                getCategoriesUseCase.fetchSiteCategories(siteModel)
+            } else
+                if (_uiState.value is Loading)
+                    _uiState.postValue(NoConnection(::fetchCategoriesFromNetwork))
```
Minor 🔍 : Since the VM is extending from the `ScopedViewModel(bgDispatcher)`, it seems all launches are running on the background thread, which requires us to always [postValue](https://developer.android.com/reference/android/arch/lifecycle/MutableLiveData#postvalue) to update the UI state. I wonder if we can move statements that do need to run on the background thread outside of the `launch` block. 🤔

AjeshRPai replied:
@ashiagr Nice catch 🙌 . I believe the UseCase functions are the only ones that need to be called in a background thread . Keeping those inside the launch block, We can move the other statements outside. Also, I will refactor the statements which are updating the UI state in the main thread to use [setValue](https://developer.android.com/reference/android/arch/lifecycle/MutableLiveData#setvalue) instead of PostValue.

AjeshRPai replied:
Have done the necessary changes in https://github.com/wordpress-mobile/WordPress-Android/pull/15552/commits/94d2046ef696f1b1cba318c0c0082dc37a04aab1 . Please have a look 👀

ashiagr commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/categories/CategoriesListViewModel.kt:
```diff
@@ -1,13 +1,131 @@
+        _uiState.postValue(Loading)
+        launch {
+            val siteCategories = getCategoriesUseCase.getSiteCategories(siteModel)
+            if (siteCategories.isNotEmpty())
+                _uiState.postValue(Content(siteCategories))
```
Super Minor 🔍 : We might want to keep control-flow statements single line if they are short and surrounded by curly braces if they're multiline. 

`if (siteCategories.isNotEmpty()) _uiState.postValue(Content(siteCategories))`

AjeshRPai replied:
Makes sense 👍.  Will follow this convention and update the statements .

AjeshRPai replied:
Have changed the Control flow statements in 94d2046. 👍

ashiagr commented on WordPress/src/test/java/org/wordpress/android/ui/prefs/categories/CategoriesListViewModelTest.kt:
```diff
@@ -1,25 +1,134 @@
+        whenever(getCategoriesUseCase.getSiteCategories(siteModel)).thenReturn(mock())
+        viewModel.start(siteModel)
+        assertTrue(uiStates.first() is Loading)
+        assertTrue(uiStates.last() is Content)
```
Suggestion 💡: It might be sufficient to just check for the `Content` here as there's a separate test case that checks for the `Loading` state for the same action (ideally, we try to have a single assertion for a test case). 

The same convention can be applied to the below test cases where we can remove the `Loading` state assertion.

`when no items are present in db and no internet available then no connection error should be displayed`
`when no items are present in db and internet is available should display list of items from network`

AjeshRPai replied:
Yeah, @ashiagr this makes sense.

AjeshRPai replied:
@ashiagr  Have done the necessary changes to assertion statements in https://github.com/wordpress-mobile/WordPress-Android/pull/15552/commits/670347820970d241370312e8a30ddd3501fca1f1

ashiagr commented on WordPress/src/test/java/org/wordpress/android/ui/prefs/categories/CategoriesListViewModelTest.kt:
```diff
@@ -1,25 +1,134 @@
+    }
+
+    @Test
+    fun `when no items are present in db and no internet available then no connection error should be displayed`() {
```
Suggestion 💡: Looking at the `when`-`then` clauses in your test statements, it seems you were trying to follow `given`-`when`-`then` format for your test statements.  In that case, I take the opportunity to elaborate that
- `when` clause is usually used for specifying an action and `given` clause for specifying pre-conditions.
- Each clause is separated with a comma (just like you did in the first two test cases). We might want to rephrase the statement to: `given no items in db with no internet, when vm starts, then no connection error is displayed`

The same convention can be applied to other test statements.

AjeshRPai replied:
Thanks for this info and explanation @ashiagr . Great Suggestion 🙌 . I think this would improve the readability of the tests to a great extent . Will do the changes 👍

AjeshRPai replied:
@ashiagr Have updated the test statements to the format of `given-when-then` format in 6703478

ashiagr commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/categories/CategoriesListFragment.kt:
```diff
@@ -4,30 +4,74 @@ import android.os.Bundle
+        super.onSaveInstanceState(outState)
+    }
+
+    override fun onStart() {
+        super.onStart()
+        dispatcher.register(this)
+    }
+
+    override fun onStop() {
+        dispatcher.unregister(this)
+        super.onStop()
+    }
+
+    @SuppressWarnings("unused")
+    @Subscribe(threadMode = MAIN)
+    fun onTaxonomyChanged(event: OnTaxonomyChanged) {
+        viewModel.onTaxonomyChanged(event)
```
Minor 🔍 : Considering the fragment and VM lifecycles, we might want to move this logic to the VM (utilise VM's `init` and `onCleared`).  I guess otherwise the `dispatcher` will be registered and unregistered unnecessarily every time the fragment is started and stopped. Wdyt?

AjeshRPai replied:
Yes @ashiagr, I think that's a better way to subscribe to the Events 👍 . I will move this logic to the ViewModel. Thanks for the suggestion 🙌

AjeshRPai replied:
Moved to VM in https://github.com/wordpress-mobile/WordPress-Android/pull/15552/commits/94d2046ef696f1b1cba318c0c0082dc37a04aab1  👍 .

ashiagr reviewed (changes requested):
Thank you for submitting your first PR @AjeshRPai! It's really nice to see you applying existing conventions, especially the single `UiState` pattern. ✨

I just left a few comments, suggestions, and one warning for you to consider. Hope it helps. Let me know if anything is not clear.

AjeshRPai committed:
Remove:NoCategories error State

As all posts are grouped under "Uncategorized" by default in other places, and the default category cannot be deleted,so there will be a Category By Default . NoCategories error State class is Redundant and can be removed.
Removes the Associated Strings from strings.xml as well
Removes the usage of these strings from the site_settings_categories_list_fragment as well

AjeshRPai committed:
Refactor: Actionable empty view

 ActionableEmptyView must have a title passed in the Constructor, Otherwise during initiation it will throw RunTime Exception and crash. Adds the title to fix the crash.

AjeshRPai committed:
Refactor: CategoriesListViewModel changes

Refactors the main thread statements to outside of the launch block
Refactors to use the SetValue to update the livedata in Main thread
Updates Control-flow statements to single line if they are short and surrounded by curly braces if they're multiline.

AjeshRPai committed:
Refactor: Moves dispatcher dependency to ViewModel

Moves the Dispatcher as a dependency to CategoriesListViewModel
Moves the subscription logic to ViewModel so that subscription doesn't happen everytime fragment is started and stopped
Adds dispatcher mock dependency to constructor in CategoriesListViewModelTest

AjeshRPai committed:
Refactor: Test statements and Assertion

Refactor: Rephrase the Test Statements in CategoriesListViewModelTest to follow given-when-then format
Removes the Unnecessary assertions in Tests from CategoriesListViewModelTest

AjeshRPai committed:
Fixes : Detekt Issue of multiline

AjeshRPai committed:
Fixes : Lint issues in Tests

ashiagr reviewed (approved):
Thank you for all the changes @AjeshRPai! 

I'll shortly initiate a merge to `develop`.

AjeshRPai commented:
@ashiagr  Thanks so much for the review 👍🙌
